x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -334,8 +342,14 @@ private void loadApiKeyAndValidateCredentials(ThreadContext ctx, ApiKeyCredentia
                         AuthenticationResult.unsuccessful("unable to find apikey with id " + credentials.getId(), null));
```

Comment:
You are right. It is better to confine the special handling in a single method. Updated as you suggested with one modification: instead of calling `listener.onResponse(AuthenticationResult.unsuccessful(...))` on the `else` branch, I think it is better to call `listener.onFailure(e)` which is consistent with original logic. We might wanna change how `unsuccessful` and `terminate` are determined in the future, but I feel it deserves a separate PR.

Reply:
> instead of calling listener.onResponse(AuthenticationResult.unsuccessful(...)) on the else branch, I think it is better to call listener.onFailure(e) which is consistent with original logic

Great, it was my intention to maintain the original behavior.

> We might wanna change how unsuccessful and terminate are determined in the future, but I feel it deserves a separate PR.

Yeah, I believe we should do that.